fendermint/vm/topdown/src/proxy.rs

```diff
@@ -64,23 +68,36 @@ impl IPCProviderProxy {
 #[async_trait]
 impl ParentQueryProxy for IPCProviderProxy {
     async fn get_chain_head_height(&self) -> anyhow::Result<BlockHeight> {
-        let height = self.ipc_provider.chain_head(&self.parent_subnet).await?;
-        Ok(height as BlockHeight)
+        emit_event_with_latency(&self.parent_subnet.to_string(), "chain_head", || async {
+            let height = self.ipc_provider.chain_head(&self.parent_subnet).await?;
+            Ok(height as BlockHeight)
+        })
+        .await
     }
 
     /// Get the genesis epoch of the child subnet, i.e. the epoch that the subnet was created in
     /// the parent subnet.
     async fn get_genesis_epoch(&self) -> anyhow::Result<BlockHeight> {
-        let height = self.ipc_provider.genesis_epoch(&self.child_subnet).await?;
-        Ok(height as BlockHeight)
+        emit_event_with_latency(&self.parent_subnet.to_string(), "genesis_epoch", || async {
+            let height = self.ipc_provider.genesis_epoch(&self.child_subnet).await?;
+            Ok(height as BlockHeight)
+        })
+        .await
     }
 
     /// Getting the block hash at the target height.
     #[instrument(skip(self))]
     async fn get_block_hash(&self, height: BlockHeight) -> anyhow::Result<GetBlockHashResult> {
-        self.ipc_provider
-            .get_block_hash(&self.parent_subnet, height as ChainEpoch)
-            .await
+        emit_event_with_latency(
```

Comment:
I kind of feel this is a bit intrusive? It's forcing the app to emit event and install the crates in monitor?

Reply:
I am not sure I follow. Could you please elaborate?

Reply:
I mean this changes the implementation of measure latency by default. Now, if I composite the current struct, `emit_event_with_latency` is called, maybe I want to measure the whole process?

 In java or node, we can use decorators. Here, maybe in the caller we use sth like at the outest level:
```rust
emit_event_with_latency(finality.get_genesis_epoch(), ...).await
```
so that we can easily remove it.

Reply:
Yeah, I was considering to have another struct (implementing the same trait) that just wraps this one and measures the latency and emits the event. Would you be happier with that?

Reply:
sounds good, my main concern is the measurement stays with this implementation but cannot easily remove it.